source/enum_set.h

```diff
@@ -12,195 +12,249 @@
-  bool ContainsWord(uint32_t word) const {
-    // We shouldn't call Overflow() since this is a const method.
-    if (auto bits = AsMask(word)) {
-      return (mask_ & bits) != 0;
-    } else if (auto overflow = overflow_.get()) {
-      return overflow->find(word) != overflow->end();
+  // Returns true if the set is holds no values.
+  bool IsEmpty() const { return buckets.size() == 0; }
+
+  // Returns true if `lhs` and `rhs` hold the exact same values.
+  friend bool operator==(const EnumSet& lhs, const EnumSet& rhs) {
+    if (lhs.buckets.size() != rhs.buckets.size()) {
+      return false;
     }
-    // The word is large, but the set doesn't have large members, so
-    // it doesn't have an overflow set.
-    return false;
+
+    for (size_t i = 0; i < lhs.buckets.size(); i++) {
+      if (rhs.buckets[i].start != lhs.buckets[i].start ||
+          rhs.buckets[i].data != lhs.buckets[i].data) {
+        return false;
+      }
+    }
+    return true;
   }
 
-  // Returns the enum value as a uint32_t.
-  uint32_t ToWord(EnumType value) const {
-    static_assert(sizeof(EnumType) <= sizeof(uint32_t),
-                  "EnumType must statically castable to uint32_t");
-    return static_cast<uint32_t>(value);
+  // Returns true if `lhs` and `rhs` hold at least 1 different value.
+  friend bool operator!=(const EnumSet& lhs, const EnumSet& rhs) {
+    return !(lhs == rhs);
   }
 
-  // Determines whether the given enum value can be represented
-  // as a bit in a uint64_t mask. If so, then returns that mask bit.
-  // Otherwise, returns 0.
-  uint64_t AsMask(uint32_t word) const {
-    if (word > 63) return 0;
-    return uint64_t(1) << word;
+  // Returns true if this set contains at least one value contained in `in_set`.
+  // Note: If `in_set` is empty, this function returns true.
+  bool HasAnyOf(const EnumSet<T>& in_set) const {
+    if (in_set.IsEmpty()) {
+      return true;
+    }
+
+    for (auto& lhs_bucket : buckets) {
+      for (auto& rhs_bucket : in_set.buckets) {
+        if (lhs_bucket.start != rhs_bucket.start) {
+          continue;
+        }
+
+        if (lhs_bucket.data & rhs_bucket.data) {
+          return true;
+        }
+      }
+    }
+    return false;
   }
 
-  // Ensures that overflow_set_ references a set.  A new empty set is
-  // allocated if one doesn't exist yet.  Returns overflow_set_.
-  OverflowSetType& Overflow() {
-    if (overflow_.get() == nullptr) {
-      overflow_ = MakeUnique<OverflowSetType>();
+ private:
+  // Storage for the buckets.
+  std::vector<Bucket> buckets;
+
+  // For a given enum `value`, finds the bucket index that could contain this
+  // value. If no such bucket is found, the index at which the new bucket should
+  // be inserted is returned.
+  size_t find_bucket_for_value(T value) const {
+    // Set is empty, insert at 0.
+    if (buckets.size() == 0) {
+      return 0;
+    }
+
+    size_t index = std::min(buckets.size() - 1, compute_bucket_index(value));
```

Comment:
`index` is not the actual index of the bucket. It represents something else. The name `compute_bucket_index` does not help either. From the names, I would assume we should just be returning `index` right away.

Now from looking at the code, I would expect that `compute_bucket_index` return largest index that could reference the bucket that contains, but you at line 230, you start looking for the bucket for value by increasing `index`. I would expect this to fail if elements are not added in sorted order.

Reply:
Buckets are sorted, that's why I can compute this index which is a perfect fit for the perfect case (enum indices are scaling linearly without gaps).
If the indices jump (like we have with capabilities), the 2nd bucket might start at value 4000, while the theoretical index is 4000/64.

For the look-left, it happens when the insertion order is reversed:
start by value 4000 -> bucket[0] is 4000.
Then, bucket 2000 -> need to scan from 4000 to the left.

I had this names `compute_theoretical_bucket_index` before, would that be better?

Reply:
I missed lines 242-243. That is where you search backwards.